### PR TITLE
Add GitHub workflow to build Docker containers

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

@geoffrey4444 once #5373 works we can add ARM to the list of platforms here to automatically build and push both architectures to DockerHub.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: .github/workflows/BuildDockerContainer.yaml
Comment (knelli2) [inline .github/workflows/BuildDockerContainer.yaml:50]: Is there a way to have this wait for another approval before pushing? Sort of like our releases? Just so we can inspect things if need be

**Review by knelli2 — COMMENTED** [review]
We'll need to be careful when we run this and how exactly we set up automatic builds, but this'll be a good start towards that.

File: .github/workflows/BuildDockerContainer.yaml
Comment (nilsvu) [reply .github/workflows/BuildDockerContainer.yaml:50]: Waiting for another approval between build and push would be more complicated because it needs to be split into separate jobs. See the `environment` option in line 21. We can configure protection rules for the environment, e.g. that it requires an approval, but it applies to the whole job.

I intend to use this on a branch to push to my own Dockerhub (or some other staging area) by setting the `image_name` to something like `nilsvu/spectre`. Then I can test it before pushing to `sxscollaboration/spectre` manually.

Comment (knelli2) [reply .github/workflows/BuildDockerContainer.yaml:50]: Do you intend to use this workflow on develop then (in addition to your fork)?

Comment (nilsvu) [reply .github/workflows/BuildDockerContainer.yaml:50]: Probably not? Usually we want to push the container to a staging area before merging a PR.

Comment (knelli2) [reply .github/workflows/BuildDockerContainer.yaml:50]: If we don't plan on using this on develop really, is there a way to only add this to your fork?

Comment (nilsvu) [reply .github/workflows/BuildDockerContainer.yaml:50]: No I think adding it to the repo is the right thing to do. It allows anyone of us to trigger the build from the PR branch. And it allows to run the tests over the new container (see new commit).

**Review by nilsdeppe — APPROVED** [review]

**Review by knelli2 — APPROVED** [review]

**Pull request merged**
